js/msp.js

```diff
@@ -160,48 +274,60 @@ var MSP = {
     },
 
     send_message: function (code, data, callback_sent, callback_msp) {
-        var bufferOut,
-            bufView,
-            i;
-
-        // always reserve 6 bytes for protocol overhead !
-        if (data) {
-            var size = data.length + 6,
-                checksum;
-
-            bufferOut = new ArrayBuffer(size);
-            bufView = new Uint8Array(bufferOut);
-
-            bufView[0] = 36; // $
-            bufView[1] = 77; // M
-            bufView[2] = 60; // <
-            bufView[3] = data.length;
-            bufView[4] = code;
-
-            checksum = bufView[3] ^ bufView[4];
-
-            for (i = 0; i < data.length; i++) {
-                bufView[i + 5] = data[i];
+        var payloadLength = data && data.length ? data.length : 0;
+        var length;
+        var buffer;
+        var view;
+        var checksum;
+        var ii;
+
+        switch (this.protocolVersion) {
+            case this.constants.PROTOCOL_V1:
+                length = payloadLength + 6;
+                buffer = new ArrayBuffer(length);
+                view = new Uint8Array(buffer);
+                view[0] = this.symbols.BEGIN;
+                view[1] = this.symbols.PROTO_V1;
+                view[2] = this.symbols.TO_MWC;
+                view[3] = payloadLength;
+                view[4] = code;
+
+                checksum = view[3] ^ view[4];
+                for (ii = 0; ii < payloadLength; ii++) {
+                    view[ii + 5] = data[ii];
+                    checksum ^= data[ii];
+                }
+                view[length-1] = checksum;
+                break;
+            case this.constants.PROTOCOL_V2:
+                length = payloadLength + 9;
+                buffer = new ArrayBuffer(length);
+                view = new Uint8Array(buffer);
+                view[0] = this.symbols.BEGIN;
+                view[1] = this.symbols.PROTO_V2;
+                view[2] = this.symbols.TO_MWC;
+                view[3] = 0; // flag: reserved, set to 0
+                view[4] = code & 0xFF;  // code lower byte
+                view[5] = (code & 0xFF00) >> 8; // code upper byte
+                view[6] = payloadLength & 0xFF; // payloadLength lower byte
+                view[7] = (payloadLength & 0xFF00) >> 8; // payloadLength upper byte
+                for (var ii = 0; ii < payloadLength; ii++) {
+                    view[7+ii] = data[ii];
```

Comment:
This overwrites payloadLength upper byte. Should be 
```
view[8+ii] = data[ii];
```

Reply:
@digitalentity Nice catch! I'll submit the requested changes in a few minutes. There are also a few style issues I've noticed since I submitted the PR that I plan to address. It was late when I finished writing that code, don't judge me :stuck_out_tongue_winking_eye:

Reply:
No worries. I suspect there is another bug with MSP payload serialization. Debugging.

Reply:
@digitalentity Just pushed the changes for version detection and those style fixes. I don't have time to run any exhaustive tests right now, so please let me know if you find any other problems and I'll look into it.